File: Meet/Meet/Me/PhotosAlbumViewController.m
//
//  PhotosAlbumViewController.m
//  Meet
//
//  Created by Zhang on 20/09/2016.
//  Copyright © 2016 Meet. All rights reserved.
//

#import "PhotosAlbumViewController.h"
#import "Meet-Swift.h"

@interface PhotosAlbumViewController ()

@end

@implementation PhotosAlbumViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (instancetype)initWithMaxImagesCount:(NSInteger)maxImagesCount delegate:(id<TZImagePickerControllerDelegate>)delegate {
    PhotosAlbumCollectionView *_albumPickerVc = [[PhotosAlbumCollectionView alloc] init];
    _albumPickerVc.block = ^(UIImage *image){
        if (self.block != nil) {
            self.block(image);
        }
    };
    self = [super initWithRootViewController:_albumPickerVc];
    if (self) {
        self.maxImagesCount = maxImagesCount > 0 ? maxImagesCount : 9; // Default is 9 / 默认最大可选9张图片
        self.pickerDelegate = delegate;
        self.selectedModels = [NSMutableArray array];
        
        // Allow user picking original photo and video, you also can set No after this method
        // 默认准许用户选择原图和视频, 你也可以在这个方法后置为NO
        self.allowPickingOriginalPhoto = YES;
        self.allowPickingVideo = YES;
        self.allowPickingImage = YES;
        self.allowTakePicture = YES;
        self.timeout = 15;
        self.photoWidth = 828.0;
        self.photoPreviewMaxWidth = 600;
        self.sortAscendingByModificationDate = YES;
        self.autoDismiss = YES;
        self.barItemTextFont = [UIFont systemFontOfSize:15];
        self.barItemTextColor = [UIColor whiteColor];
        [self initImageAlbum];
    }
    return self;
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


@end

@interface PhotosAlbumCollectionView()<FusumaDelegate>

@end

@implementation PhotosAlbumCollectionView

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    FusumaViewController *fusuma = [[FusumaViewController alloc] init];
    fusuma.delegate = self;
    TZAlbumModel *model = [self.albumArr objectAtIndex:indexPath.row];
    fusuma.images = model.result;
    [self.navigationController pushViewController:fusuma animated:YES];
}


- (void)fusumaImageSelected:(UIImage * _Nonnull)image
{
    if (self.block != nil) {
        self.block(image);
    }
}

- (void)fusumaDismissedWithImage:(UIImage * _Nonnull)image
{
    
}

- (void)fusumaVideoCompletedWithFileURL:(NSURL * _Nonnull)fileURL
{
    
}

- (void)fusumaCameraRollUnauthorized
{
    
}


- (void)fusumaClosed
{
    
}

@end

File: Meet/Meet/Me/PhotosAlbumViewController.swift
//
//  PhotosAlbumViewController.swift
//  PhotoAlbum
//
//  Created by Zhang on 8/27/16.
//  Copyright © 2016 Zhang. All rights reserved.
//

import UIKit
import Photos

@objc public protocol PhotosAlbumViewControllerDelegate : class {
    func albumImageSelected(_ image: UIImage)
    @objc optional func albumImageDismissedWithImage(_ image: UIImage)
    func albumImageVideoCompleted(withFileURL fileURL: URL)
    func albumImageCameraRollUnauthorized()
    @objc optional func albumImageClosed()
}


class PhotosAlbumViewController: UIViewController {

    var tableView:UITableView!
    var listArray:NSMutableArray = NSMutableArray()
    
    internal weak var delegate: PhotosAlbumViewControllerDelegate? = nil
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.configTableView()
        self.setUpTableView()
        self.navigationItemWithLineAndWihteColor()
        self.title = "选择相册"
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage.init(named: "me_dismissBlack")?.withRenderingMode(.alwaysOriginal), style: .plain, target: self, action: #selector(PhotosAlbumViewController.dismissView))
        // Do any additional setup after loading the view.
    }
    
    func dismissView() {
        self.dismiss(animated: true) { 
            
        }
    }

    func configTableView() {
        let manager = TZImageManager()
        manager.getAllAlbums(false, allowPickingImage: true) { (model) in
            self.listArray.addObjects(from: model!)
        }
    }
    
    func setUpTableView() {
        if tableView == nil {
            tableView = UITableView(frame: CGRect(x: 0, y: 0, width: ScreenWidth, height: ScreenHeight - 64), style: .plain)
            tableView.delegate = self
            tableView.dataSource = self
            tableView.rowHeight = 90
            tableView.separatorStyle = .none
            tableView.register(TZAlbumCell.self, forCellReuseIdentifier: "PhotosAlbum")
            self.view.addSubview(tableView)
        }else{
           tableView.reloadData()
        }
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

extension PhotosAlbumViewController : UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let fusuma = FusumaViewController()
        fusuma.delegate = self
        fusuma.images = (self.listArray[(indexPath as NSIndexPath).row] as! TZAlbumModel).result as! PHFetchResult
        
        self.navigationController?.pushViewController(fusuma, animated: true)
    }
}

extension PhotosAlbumViewController : UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.listArray.count
    }
    
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdef = "PhotosAlbum"
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdef,for: indexPath) as! TZAlbumCell
        let model = self.listArray[(indexPath as NSIndexPath).row] as! TZAlbumModel
        cell.model = model
        cell.selectionStyle = .none
        return cell
    }
}



extension PhotosAlbumViewController : PHPhotoLibraryChangeObserver {
    func photoLibraryDidChange(_ changeInstance: PHChange) {
        
    }
}


extension PhotosAlbumViewController : FusumaDelegate {
    func fusumaImageSelected(_ image: UIImage) {
        self.delegate?.albumImageSelected(image)
    }
    func fusumaDismissedWithImage(_ image: UIImage) {
        self.delegate?.albumImageDismissedWithImage!(image)
    }
    func fusumaVideoCompleted(withFileURL fileURL: URL) {
        self.delegate?.albumImageVideoCompleted(withFileURL: fileURL)
    }
    func fusumaCameraRollUnauthorized() {
        self.delegate?.albumImageCameraRollUnauthorized()
    }
    
    func fusumaClosed() {
        self.delegate?.albumImageClosed!()
    }
}
